### PR TITLE
Fix readability & caching on CRM_Contact_BAO_Relationship::isInheritedMembershipInvalidated

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
Mostly code readability but also improves caching

Before
----------------------------------------
Sql query & some hard-to-read & unnecessary wrangling 

After
----------------------------------------
More readable. Introduction of a cached function improves performance

Technical Details
----------------------------------------
@seamuslee001 looking at this I was going to use the 'fields' cache but of course the name was wrong so I figured adding a generic 'metadata' cache would cover a lot of uses (including probably what 'fields' currently is....

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Contact/BAO/Relationship.php [hunk 1]
```diff
@@ -2321,22 +2321,18 @@ private static function isRelationshipTypeCurrentEmployer(int $existingTypeID):
...
-      $relTypeId = $dao->relationship_type_id;
-    }
-    $relTypeIds = explode(CRM_Core_DAO::VALUE_SEPARATOR, $relTypeId);
-    $isDeletable = in_array($values[$cid]['relationshipTypeId'], $relTypeIds
-      //CRM-16300 check if owner membership exist for related membership
-      ) && !empty($membershipValues['owner_membership_id']) && !empty($values[$mainRelatedContactId]['memberships'][$membershipValues['owner_membership_id']]);
-    return [$relTypeId, $isDeletable];
+    $membershipType = CRM_Member_BAO_MembershipType::getMembershipType($membershipValues['membership_type_id']);
+    $relTypeIds = $membershipType['relationship_type_id'];
+    $membshipInheritedFrom = $membershipValues['owner_membership_id'] ?? NULL;
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Relationship.php]: I just broke the IF into 2 because  it was hard to read & a later fix would be more nuanced on the  second part

**Comment by eileenmcnaughton** [comment]
Refactoring relates to #14410

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this failed on something to do with the cache - possibly I need to add a clear. Before I dig - do you agree with adding this cache?

**Comment by seamuslee001** [comment]
I think that makes sense

**Comment by eileenmcnaughton** [comment]
@seamuslee001 looks like it was the 'clear' thing - passing now

**Comment by eileenmcnaughton** [comment]
Actually I just pushed a tweak so they are re-running

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this is passing now

**Comment by seamuslee001** [comment]
I think the code looks fine to me and i think we should have enough test coverage in this are. Merging

**Pull request merged**